### PR TITLE
Completed express-notes-api challenge.

### PR DESCRIPTION
**Pull request by brandon-moy**
Get Demo:

[get demo.webm](https://user-images.githubusercontent.com/109567499/196049208-1b745336-1be7-41fe-b8f2-50722c03a4a9.webm)

Post Demo:

[post demo.webm](https://user-images.githubusercontent.com/109567499/196049212-ae5ce5b1-a2f6-4feb-a61a-649102fdaa9b.webm)

Put Demo:

[put demo.webm](https://user-images.githubusercontent.com/109567499/196049222-90eb34b3-44de-4f8f-b22c-024dd69a70a5.webm)

Delete Demo:

[delete demo.webm](https://user-images.githubusercontent.com/109567499/196049224-96ea4394-392f-4a19-a687-b5f6350c49aa.webm)

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    } else {
+      res.status(201);
+      if (type === 'post') {
+        res.json(note);
+      } else {
+        res.send();
+      }
+    }
+  });
+};
```
Comment (thebearingedge) [inline express-notes-api/index.js:41]: This is a little bit too fancy. The code is hard to follow if you have to trace where the `res` object is going and what's happening to it. Try this instead.

```js
const updateDataJSON = (data, callback) => {
  const updatedJSON = JSON.stringify(data, null, 2);
  fs.writeFile('./data.json', updatedJSON, callback);
});
```

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    res.json({
+      error: 'id must be a positive integer'
+    });
+  } else if (data.notes[req.params.id] === undefined) {
+    res.status(404);
+    res.json({
+      error: `cannot find note with id ${req.params.id}`
+    });
+  }
+};
```
Comment (thebearingedge) [inline express-notes-api/index.js:21]: Rework this to just answer the question, is the `id` valid

```js
const idCheck = id => {
  if (!(id in data.notes)) return 'not found';
  if (!Number.isInteger(Number(id))) return 'invalid';
  return null;
}
```

**Review by thebearingedge — CHANGES_REQUESTED** [review]

File: express-notes-api/index.js (hunk 1 above)
Comment (brandon-moy) [reply express-notes-api/index.js]: Got it. I had written the function this way to try and reduce the lines of code in the overall document. Is it better to keep the `res` object out of these kinds of functions for clarity then?

image

image

File: express-notes-api/index.js (hunk 2 above)
Comment (brandon-moy) [reply express-notes-api/index.js]: I also made the adjustment to this function, but then I would need to add the conditionals into each HTTP request method.

image

image

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+const updateDataJSON = data => {
+  const updatedJSON = JSON.stringify(data, null, 2);
+  fs.writeFile('./data.json', updatedJSON, err => {
+    if (err) {
+      console.error(err);
+      return 'error';
+    }
+  });
+};
```
Comment (thebearingedge) [inline express-notes-api/index.js:23]: this function needs to take a callback so you can do your error handling / responding in your routes. you can't return from an asynchronous callback. don't write code you can't confirm works as intended.
```js
const updateDataJSON = (data, callback) => {
  const updatedJSON = JSON.stringify(data, null, 2);
  fs.writeFile('./data.json', updatedJSON, callback);
};
```

```js
updateJSON(data, err => {
  if (err) {
    // send 500
  } else {
    // send 2xx
  }
});
```

**Review by thebearingedge — CHANGES_REQUESTED** [review]

File: express-notes-api/index.js (hunk 3 above)
Comment (brandon-moy) [reply express-notes-api/index.js:23]: I think I get it now. I couldn't connect the relation of the callback being where the error and response is handled and having that passed through as an argument when the `updateDataJSON` function was being called

image

image

**Review by thebearingedge — APPROVED** [review]
💯
